make/containers.mk

```diff
@@ -0,0 +1,90 @@
+BUILDER_IMAGE=docker.io/golang:1.17-bullseye
+
+# set to "DYNAMIC" to use a dynamic base image
+BASE_IMAGE_TYPE:=STATIC
+
+BASE_IMAGE_controller-linux-amd64:=$($(BASE_IMAGE_TYPE)_BASE_IMAGE_amd64)
+BASE_IMAGE_controller-linux-arm64:=$($(BASE_IMAGE_TYPE)_BASE_IMAGE_arm64)
+BASE_IMAGE_controller-linux-s390x:=$($(BASE_IMAGE_TYPE)_BASE_IMAGE_s390x)
+BASE_IMAGE_controller-linux-ppc64le:=$($(BASE_IMAGE_TYPE)_BASE_IMAGE_ppc64le)
+BASE_IMAGE_controller-linux-armv7:=$($(BASE_IMAGE_TYPE)_BASE_IMAGE_armv7)
+
+BASE_IMAGE_webhook-linux-amd64:=$($(BASE_IMAGE_TYPE)_BASE_IMAGE_amd64)
+BASE_IMAGE_webhook-linux-arm64:=$($(BASE_IMAGE_TYPE)_BASE_IMAGE_arm64)
+BASE_IMAGE_webhook-linux-s390x:=$($(BASE_IMAGE_TYPE)_BASE_IMAGE_s390x)
+BASE_IMAGE_webhook-linux-ppc64le:=$($(BASE_IMAGE_TYPE)_BASE_IMAGE_ppc64le)
+BASE_IMAGE_webhook-linux-armv7:=$($(BASE_IMAGE_TYPE)_BASE_IMAGE_armv7)
+
+BASE_IMAGE_acmesolver-linux-amd64:=$($(BASE_IMAGE_TYPE)_BASE_IMAGE_amd64)
+BASE_IMAGE_acmesolver-linux-arm64:=$($(BASE_IMAGE_TYPE)_BASE_IMAGE_arm64)
+BASE_IMAGE_acmesolver-linux-s390x:=$($(BASE_IMAGE_TYPE)_BASE_IMAGE_s390x)
+BASE_IMAGE_acmesolver-linux-ppc64le:=$($(BASE_IMAGE_TYPE)_BASE_IMAGE_ppc64le)
+BASE_IMAGE_acmesolver-linux-armv7:=$($(BASE_IMAGE_TYPE)_BASE_IMAGE_armv7)
+
+BASE_IMAGE_cainjector-linux-amd64:=$($(BASE_IMAGE_TYPE)_BASE_IMAGE_amd64)
+BASE_IMAGE_cainjector-linux-arm64:=$($(BASE_IMAGE_TYPE)_BASE_IMAGE_arm64)
+BASE_IMAGE_cainjector-linux-s390x:=$($(BASE_IMAGE_TYPE)_BASE_IMAGE_s390x)
+BASE_IMAGE_cainjector-linux-ppc64le:=$($(BASE_IMAGE_TYPE)_BASE_IMAGE_ppc64le)
+BASE_IMAGE_cainjector-linux-armv7:=$($(BASE_IMAGE_TYPE)_BASE_IMAGE_armv7)
+
+bin/containers:
+	@mkdir -p $@
+
+.PHONY: all-containers
+all-containers: cert-manager-controller-linux cert-manager-webhook-linux cert-manager-acmesolver-linux cert-manager-cainjector-linux
+
+.PHONY: cert-manager-controller-linux
+cert-manager-controller-linux: bin/containers/cert-manager-controller-linux-amd64.tar.gz bin/containers/cert-manager-controller-linux-arm64.tar.gz bin/containers/cert-manager-controller-linux-s390x.tar.gz bin/containers/cert-manager-controller-linux-ppc64le.tar.gz bin/containers/cert-manager-controller-linux-armv7.tar.gz
+
+bin/containers/cert-manager-controller-linux-%.tar.gz: bin/controller-linux-% make/Containerfile.controller | bin/containers
```

Comment:
So this will be re-run every time the built binary has changed or the Containerfile has changed. What if TAG, BASE_IMAGE or GOARGS has changed? (For example, with Bazel we use [stamping](https://github.com/jetstack/cert-manager/tree/master/build#stamping) to ensure that targets will re-run i.e if a different image tag is set)

Reply:
Good point. I can't think of any "good" way to rebuild the target on env var changes.

I think it is not a deal-breaker though 😄   It wouldn't be a problem on CI anyways, but maybe it would become problematic locally.